string_indexing/suffix_array.py

```diff
@@ -63,6 +66,223 @@ def induced_sorting(text, n):
+    if SL.count(Category.S) == 1:
+      return list(reversed(range(1, N + 1)))
+
+    if SL[-1] == Category.S:
+      sortedS = _sort_category(s, SL, Category.S)
+      groups = A.groups.copy()
+      _move_to_group_end(
+          reversed(sortedS[1:]),
+          lambda x: True, False, A, groups)
+      _move_to_group_end(
+          map(lambda i : A.tab[i] - 1, range(1, len(A.tab))),
+          lambda x: SL[x] == Category.L, True, A, groups)
+    elif SL[-1] == Category.L:
+      sortedL = _sort_category(s, SL, Category.L)
+      groups = A.groups.copy()
+      _move_to_group_end(sortedL[1:], lambda x: True, True, A, groups)
+      _move_to_group_end(
+          map(lambda i : A.tab[i] - 1, reversed(range(1, len(A.tab)))),
+          lambda x: SL[x] == Category.S, False, A, groups)
+    else:
+      raise Exception("fail")
+
+    return A.tab[1:]
+
+  def _sort_category(s, SL, category):
+    suff_by_dist, categ_substr = _prepare_substr_lists(
+        _bucket_array(s), SL, category)
+
+    # sort categ_substr
+    for i in range(1, len(suff_by_dist)):
+      groups = suff_by_dist[i].sorted_groups()
+      for g in groups if category == Category.S else reversed(groups):
+        to_split = collections.defaultdict(list)
+        for j in suff_by_dist[i].tab[g[0]:g[1]+1]:
+          x = j-i
+          g_id = categ_substr.group_id[categ_substr.rev[x]]
+          to_split[g_id].append(x)
+
+        for g_id in to_split:
+          vals = to_split[g_id]
+          if category == Category.S:
+            categ_substr.split_group_before(vals)
+            categ_substr.reverse_group(
+                categ_substr.group_id[categ_substr.rev[vals[0]]])
+          elif category == Category.L:
+            categ_substr.split_group_after(vals)
+            categ_substr.reverse_group(
+                categ_substr.group_id[categ_substr.rev[vals[0]]])
+
+    # map Ssubstr to ints
+    categ_substr.refresh_group_ids()
+    rec_s, m = [-1], { -1 : -1 }
+    for i in range(1, len(s)):
+      if SL[i] == category:
+        m[len(m)] = i
+        rec_s.append(categ_substr.group_id[categ_substr.rev[i]])
+
+    # recursive call and undo mapping
+    return [ m[i] for i in _small_large(rec_s) ]
+
+  def _move_to_group_end(vals, condition, front, A, groups):
+    for v in vals:
+      if not condition(v):
+        continue
+      idx = A.rev[v]
+      g_id = A.group_id[idx]
+      g = groups[g_id]
+
+      if front:
+        A.swap(idx, g[0])
+        groups[g_id] = (g[0]+1, g[1])
+      else:
+        A.swap(idx, g[1])
+        groups[g_id] = (g[0], g[1]-1)
+
+  def _prepare_substr_lists(A, SL, category):
+    categDist = _calc_category_dist(SL, category)
+    suff_by_dist = [ GroupedArray() for _ in range (max(categDist)+1) ]
+    category_substr = GroupedArray()
+
+    for g in A.sorted_groups():
+      to_add = []
+      for i in range(g[0], g[1]+1):
+        x = A.tab[i]
+        if SL[x] == category:
+          to_add.append(x)
```

Comment:
I guess we can change it to:
category_substr.append_group([A.tab[i] for SL[A.tab[i]] == category for i in range(g[0], g[1]+1)])